### PR TITLE
#12228 Forbid NUL in HTTP header values

### LINKED ISSUE #12228: Twisted incorrectly allows NUL in header values
# The incorrect behavior
From RFC 9110:
> Field values containing CR, LF, or NUL characters are invalid and dangerous, due to the varying ways that implementations might parse and interpret those characters; a recipient of CR, LF, or NUL within a field value MUST either reject the message or replace each of those characters with SP before further processing or forwarding of that message.

# How to cause this behavior

0. Install Twisted
1. Run the following script (it just echos back the parsed components of each received request in base64, inside a JSON object):
```python3
from base64 import b64encode

from twisted.web import server, resource
from twisted.internet import reactor

class TheResource(resource.Resource):
    isLeaf = True
    def __getattribute__(self, key: str):
        if key.startswith("render_"):
            return self._respond
        return super().__getattribute__(key)

    def _respond(self, request) -> bytes:
        result: bytes = b'{"headers":['
        result += b",".join(b'["' + b64encode(name) + b'","' + b64encode(value) + b'"]' for name, value in request.getAllHeaders().items())
        result += b'],"method":"' + b64encode(request.method)
        result += b'","uri":"' + b64encode(request.uri)
        result += b'","version":"' + b64encode(request.clientproto)
        result += b'","body":"' + b64encode(request.content.read())
        result += b'"}'
        return result

site = server.Site(TheResource())
reactor.listenTCP(80, site)
reactor.run()
```
2. Send it a request with a NUL byte in a header value, and extract the parsed header value from the response:
```sh
(printf 'GET / HTTP/1.1\r\nHost: whatever\r\nTest: a\x00a\r\n\r\n'; sleep 1) \
    | nc localhost 80 \
    | grep '"headers"' \
    | jq '.["headers"][1][1]' \
    | xargs echo \
    | base64 -d \
    | xxd
```
```
00000000: 6100 61                                  a.a
```

# The correct behavior I'd like to see
Twisted should either reject the request with a 400, or replace the NUL with a SP before processing.

# Testing environment
 - Operating System and Version; paste the output of these commands:
   - `Linux cb97695853e1 6.9.7-arch1-1 #1 SMP PREEMPT_DYNAMIC Fri, 28 Jun 2024 04:32:50 +0000 x86_64 GNU/Linux`
 - Twisted version [e.g. 22.2.0]
   - `24.3.0.post0`

### PR DESCRIPTION
**Pull request by twm**
## Scope and purpose

Fixes #12228.

I chose to forbid NUL values rather than coerce them to whitespace. Note that this only applies to the HTTP server — `http_headers.Headers` still permits all sorts of malformed header names and values.

## Contributor Checklist:

This process applies to *all* pull requests - no matter how small.
Have a look at [our developer documentation](https://docs.twisted.org/en/latest/core/development/dev-process.html) before submitting your Pull Request.

Below is a non-exhaustive list (as a reminder):

* The title of the PR should describe the changes and starts with the associated issue number, like “#1234 Brief description”.
* A release notes news fragment file was create in src/twisted/newsfragments/ (see: [Release notes fragments docs.](https://docs.twisted.org/en/latest/core/development/dev-process.html#release-notes-management))
* The automated tests were updated.
* Once all checks are green, request a review by leaving a comment that contains exactly the string `please review`.
  Our bot will trigger the review process, by applying the pending review label
  and requesting a review from the Twisted dev team.

### CODE REVIEW

File: src/twisted/web/newsfragments/12233.bugfix [hunk 1]
```diff
@@ -1 +1 @@
-The HTTP 1.0/1.1 server provided by twisted.web is now more picky about the first line of a request, improving compliance with RFC 9112.
+twisted.web's HTTP/1.1 server is now more picky about the first line of a request, improving compliance with RFC 9112.
```
Comment (adiroiban) [inline src/twisted/web/newsfragments/12233.bugfix]: I think that we should revert the changes to 12233 and 12235.

If these changes are needed, they should be done on the NEWS file inside the release pr https://github.com/twisted/twisted/pull/12272

**Comment by adiroiban** [comment]
I guess that it's expected to see the performance decrease.
We are now doing more work.

I would prefer to get the final release PR approved and merged, and then merge this PR

Otherwise the changes to the existing news fragments will not be found in the release

thanks

File: src/twisted/web/newsfragments/12233.bugfix (hunk 1 above)
Comment (twm) [reply src/twisted/web/newsfragments/12233.bugfix]: Absolutely, I'll take care of that when I merge forward.

**Comment by twm** [comment]
> I guess that it's expected to see the performance decrease.
> We are now doing more work.

Coercing NUL to a space could be done as part of the linear whitespace sanitization process, likely with less perf impact. That happens in the `Headers` class, though, so it'd also affect `Agent`. TBH I think that'd be fine, but it does increase the potential compatibility impact of the change.

**Review by adiroiban — APPROVED** [review]
Thanks. It looks good.

I think that we can merge it.

Having the validation inside the Header class, makes more sense, but I think that it might have a bigger performance impact, overall.
